products/llm_observability/frontend/LLMObservabilityTraceScene.tsx

```diff
@@ -189,72 +190,151 @@ function NestingGroup({ level = 0, children }: { level?: number; children: React
+        topLevelTrace,
+        item,
+        isSelected,
+    }: {
+        topLevelTrace: LLMTrace
+        item: LLMTrace | LLMTraceEvent
+        isSelected: boolean
+    }): JSX.Element => {
+        const totalCost = 'properties' in item ? item.properties.$ai_total_cost_usd : item.totalCost
+        const latency = 'properties' in item ? item.properties.$ai_latency : item.totalLatency
+        const usage = formatLLMUsage(item)
+
+        const children = [
+            isLLMTraceEvent(item) && item.properties.$ai_is_error && (
+                <LemonTag key="error-tag" type="warning">
+                    Error
+                </LemonTag>
+            ),
+            latency >= 0.01 && (
+                <LemonTag key="latency-tag" type="muted">
+                    {formatLLMLatency(latency)}
+                </LemonTag>
+            ),
+            (usage != null || totalCost != null) && (
+                <span key="usage-tag">
+                    {usage}
+                    {usage != null && totalCost != null && <span>{' / '}</span>}
+                    {totalCost != null && formatLLMCost(totalCost)}
+                </span>
+            ),
+        ]
+        const hasChildren = children.find((child) => !!child)
+
+        return (
+            <li key={item.id} className="mt-0.5">
+                <Link
+                    to={urls.llmObservabilityTrace(topLevelTrace.id, {
+                        event: item.id,
+                        timestamp: removeMilliseconds(topLevelTrace.createdAt),
+                    })}
+                    className={classNames(
+                        'flex flex-col gap-1 p-1 text-xs rounded min-h-8 justify-center hover:bg-accent-primary-highlight',
+                        isSelected && 'bg-accent-primary-highlight'
+                    )}
+                >
+                    <div className="flex flex-row items-center gap-1.5">
+                        <EventTypeTag event={item} size="small" />
+                        <span className="flex-1 truncate">{formatLLMEventTitle(item)}</span>
+                    </div>
+                    {hasChildren && (
+                        <div className="flex flex-row flex-wrap text-muted items-center gap-1.5">{children}</div>
+                    )}
+                </Link>
+            </li>
+        )
+    }
+)
+TraceNode.displayName = 'TraceNode'
+
+function RecursiveTreeDisplay({
+    tree,
+    trace,
+    selectedEventId,
 }: {
-    topLevelTrace: LLMTrace
-    item: LLMTrace | LLMTraceEvent
-    isSelected: boolean
+    tree: TraceTreeNode[]
+    trace: LLMTrace
+    selectedEventId?: string | null
 }): JSX.Element {
-    const totalCost = 'properties' in item ? item.properties.$ai_total_cost_usd : item.totalCost
-    const latency = 'properties' in item ? item.properties.$ai_latency : item.totalLatency
-    const usage = formatLLMUsage(item)
+    return (
+        <NestingGroup level={1}>
+            {tree.map(({ event, children }) => (
```

Comment:
Something mysterious I'm seeing in Max AI traces: a span _beyond_ what I can scroll to at the bottom
![Screenshot 2025-01-28 at 16 33 21](https://github.com/user-attachments/assets/f033a453-4d92-4ec8-b7e9-aaf1a0c65a83)